adds lifting-state-up content (#40)
Ada-Developers-Academy/core-unit-3: lifting-state-up/passing-down-event-handlers.md

This function should go inside the `App` component, after `studentData` is defined, and before the `return` statement.

<!-- prettier-ignore-start -->
```js
  const updateStudentData = updatedStudent => {
    const students = studentData.map(student => {
      if (student.id === updatedStudent.id) {
        return updatedStudent;
      } else {
        return student;
      }
    });

    setStudentData(students);
  }
```
<!-- prettier-ignore-end -->

| <div style="min-width:300px;"> Piece of Code </div> | Notes                                                                                                                                                                |
| --------------------------------------------------- | -------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| `const updateStudentData = `                        | We're creating a function named `updateStudentData`.                                                                                                                 |

> | `updatedStudent`                                    | This function accepts one argument: an object that holds updated student data. We expect this object to have the key `id`, at the very least.                        |

Author: I'm thinking through this comment right now as I write the forms lesson.

I think right now I have the opinion to mismatch the names so it's a little inconvenient, and force more object manipulation. Hmmmmm....

> | `const students = ;`                                | For this function, we'll create a helper array, `students` to contain the updated student data.                                                                      |
> | `studentData.map(...)`                              | We'll map over `studentData`...                                                                                                                                      |
> | `student => {...}`                                 | Inside `map`, we'll reference each item with `student`...                                                                                                            |
> | `if (student.id === updatedStudent.id) {`           | If we found the student that our `updatedStudent` is updating...                                                                                                     |

Reviewer: voice in present tense. since the example has already been shown (past would be ok, but present makes it sound more immediate)
```suggestion
| `const students = ;`                                | For this function, we create a helper array, `students` to contain the updated student data.                                                                      |
| `studentData.map(...)`                              | We map over `studentData`...                                                                                                                                      |
| `student => {...}`                                 | Inside `map`, we reference each item with `student`...                                                                                                            |
| `if (student.id === updatedStudent.id) {`           | If we find the student that our `updatedStudent` is updating...                                                                                                     |
```

> | `return updatedStudent`                             | The item we should map to our `students` array is the updated student data!                                                                                          |

Reviewer: "map to" is fine from a mathematical perspective, but just from how map works as a function, and the idea that we're building values to go into the new array, "into" sounds better there
```suggestion
| `return updatedStudent`                             | The item we should map into our `students` array is the updated student data!                                                                                          |
```

| `else { return student; }`                          | Otherwise, we should map the unchanged `student` into `students`.                                                                                                    |

> | `setStudentData(students);`                         | Ultimately, we want to update the `studentData` on state. We use our state update function, `setStudentData`, and we update it to our newly formed `students` array. |

Reviewer: there are a few places where the phrasing "on state" is used. Was this left over from the idea that there would be an actual state object (class-like React) and these values would be properties "on" that state object?
```suggestion
| `setStudentData(students);`                         | Ultimately, we want to update the `studentData` in our state. We use our state update function, `setStudentData`, and we update it to our newly formed `students` array. |
```

Our `App` now defines and holds a function that can update our `studentData`, named `updateStudentData`.

> 

Reviewer: suggest a callout about why the heck we're copying the whole array, and making a very important (but not obvious) point about how React decides whether to trigger a re-render.
```suggestion

### !callout-info

## Helping React Notice That `studentData` Changed

Notice that during the update, we made a helper array that is more or less a copy of the data we already had. For every student whose ID does not match the `updatedStudent`, we reuse the student data that was previously in our state. So why couldn't we simply find the matching student data in our existing state and update that?

<br/>

Because React wouldn't notice the change! For container types, like arrays and objects, if we update an inner value (an item in an array, or a field in an object) and then call the relevant set function, React will see that the reference to the object currently in state is the same reference as the object being passed into the set function. For performance reasons, it assumes that if the same reference is passed in to a set function, it _is_ the same object. It will not look within the array or object for changes!

<br/>

So to get React to notice that a value in our array or object has changed, we need to make a new outer reference, and copy the existing values into it. Then when we call a set function, it will see the change and trigger a re-render.

### !end-callout

```

## Send This Function to `StudentList`

Now that we've defined `updateStudentData`, imagine if other components could _use_ this function.

If other components could use this function, they'd have the ability to update the `studentData` in `App`! We want _the `Student` component_ to update the `studentData` in `App`.

> Other components _can_ use the `updateStudentData` function if they can reference it... and they can, _using props_.

Reviewer: this extra sentence might sound too redundant, but for some reason I feel like reiterating this (including reference as well)
```suggestion
Other components _can_ use the `updateStudentData` function if they can reference it... and they can, _using props_. We can pass a reference to the `updateStudentData` down to a component through its `props`!
```

First, let's change `App` so that it sends this function to `StudentList` through a new prop named `onUpdateStudent`.

<!-- prettier-ignore-start -->
```js
  return (
    <main>
      <h1>Attendance</h1>
      <StudentList
        students={studentData}
        onUpdateStudent={updateStudentData}
      ></StudentList>
    </main>
  );
```
<!-- prettier-ignore-end -->

> Let's update the PropTypes for `StudentList`. Our `StudentList` component now expects a `prop` named `onUpdateStudent`, whose value is a _function_.

Reviewer: include reference
```suggestion
Let's update the PropTypes for `StudentList`. Our `StudentList` component now expects a `prop` named `onUpdateStudent`, whose value is a _reference to a function_.
```

<!-- prettier-ignore-start -->
```js
StudentList.propTypes = {
    students: PropTypes.arrayOf(PropTypes.shape({
        id: PropTypes.number.isRequired,
        nameData: PropTypes.string.isRequired,
        emailData: PropTypes.string.isRequired,
        isPresentData: PropTypes.bool
    })),

>     onUpdateStudent: PropTypes.func

Reviewer: Since this winds up being set as required in Student later on, we should probably mark it required here as well
```suggestion
    onUpdateStudent: PropTypes.func.isRequired
```

};
```
<!-- prettier-ignore-end -->

## Send This Function to `Student`

> Now, let's send this _exact same function_ from the `StudentList` component to the `Student` component.

Reviewer: include reference
```suggestion
Now, let's send this _exact same function reference_ from the `StudentList` component to the `Student` component.
```

We access this function through `props`, and we send it to `Student` through `props`.

The keys to reading and writing this code are to:

1. Check that we're _reading_ the correct `prop` that was passed in
2. Recognize that we can use any name for any `prop` we send to `Student`

<!-- prettier-ignore-start -->
```js
const StudentList = (props) => {
    const studentComponents = props.students.map((student, index) => {
        return (
            <li key={index}>
                <Student
                    id={student.id}
                    name={student.nameData}
                    email={student.emailData}
                    isPresent={student.isPresentData}
                    onUpdate={props.onUpdateStudent}
                ></Student>
            </li>
        );
    });

    // ... return some JSX
```
<!-- prettier-ignore-end -->

> Here, we're saying that each `Student` component has a `prop` named `onUpdate`. The value of `onUpdate` is `props.onUpdateStudent`.

Reviewer: say what we expect to be there (reinforcing we have to remember what these things are supposed to be)
```suggestion
Here, we're saying that each `Student` component has a `prop` named `onUpdate`. The value of `onUpdate` will be whatever was passed in `props.onUpdateStudent`, which we expect to be a reference to our `updateStudentData` function.
```

* `Student` needs a function that updates the student data, because `isPresent` is managed in `Student`'s state. The initial student data is passed to `StudentList` and `Student` through `props`.

##### !end-options
##### !answer

* `App` needs a function that updates the student data, because the student data is managed in `App`'s state. In order to use this function, we need to pass this function to `StudentList` and `Student` through `props`.

##### !end-answer
### !end-challenge
<!-- prettier-ignore-end -->

<!-- Question 2 -->
<!-- prettier-ignore-start -->
### !challenge
* type: multiple-choice
* id: f581b6de
* title: Passing Down Event Handlers
##### !question

Which of the following options best describes this function, `updateStudentData`?

```js
  const updateStudentData = updatedStudent => {
    const students = studentData.map(student => {
      if (student.id === updatedStudent.id) {
        return updatedStudent;
      } else {
        return student;
      }
    });

    setStudentData(students);

>   }

Reviewer: ```suggestion
  };
```

```

##### !end-question
##### !options

> * This function creates an array `studentData`, which shares the same name as `studentData` on state, but it isn't the same.
> * This function creates an array `students`, which contains objects of student data. It updates `studentData` on state by invoking `setStudentData`.

Reviewer: on → in
```suggestion
* This function creates an array `studentData`, which shares the same name as `studentData` in state, but it isn't the same.
* This function creates an array `students`, which contains objects of student data. It updates `studentData` in state by invoking `setStudentData`.
```

* This function updates `studentData` by invoking `setStudentData`, and passing in the single object `updatedStudent`.

##### !end-options
##### !answer

> * This function creates an array `students`, which contains objects of student data. It updates `studentData` on state by invoking `setStudentData`.

Reviewer: on → in
```suggestion
* This function creates an array `students`, which contains objects of student data. It updates `studentData` in state by invoking `setStudentData`.
```
